### PR TITLE
fix: give concurrent streams disjoint row_states buffers in multi-CTA top-k

### LINKED ISSUE #3618: [Bug] Multi-CTA top-k row_states buffer is shared across concurrent streams (per-device cache key)
## Description

Found while root-causing #3610 (fix: #3615), filing separately since it's pre-existing and orthogonal.

The multi-CTA radix top-k entry points allocate their software-barrier state via `_get_cache_buf` with a **per-device** key, e.g. `flashinfer/sampling.py`:

```python
row_states_buffer = _get_cache_buf(
    f"top_k_renorm_probs_row_states_{probs.device}", buffer_bytes, probs.device, zero_init=True,
)
```

(same pattern for `top_k_mask_logits` and the other multi-CTA top-k entry points.)

Two streams on the same device calling the **same API** concurrently therefore share one `RadixRowState` array: both kernels' CTA groups increment/spin on the *same* `arrival_counter` words and write the same histograms. Outcomes range from wrong top-k results to a wedged barrier (one kernel observes counter values advanced by the other), independent of the #3615 fixes — that PR makes each launch's own protocol sound, but cross-launch sharing of the state words remains.

During #3615 validation, the 2-stream stress case only passes with **separate state buffers per stream** (noted in the PR's validation table); with the shared buffer it's an unsynchronized data race by inspection.

## Suggested fix directions

- Key the cache by `(device, stream)` — cheap, but buffers accumulate per stream and stream lifetimes are opaque from Python; or
- Single per-device buffer + slicing by a launch-serialized offset (ring of N state arrays, advanced per launch) — bounded memory, no stream identity needed; or
- Move state allocation into the C++ launcher behind a stream-ordered allocator (`cudaMallocAsync`) so each launch gets private state with no Python-side cache at all.

Happy to take this (have the SM120 repro rig from #3610/#3615 and a 2-stream stress harness ready); opening the issue first per contributing conventions since the right allocation strategy is a maintainer call.

### PR DESCRIPTION
**Pull request by waynehacking8**
## Description

Fixes #3618.

The multi-CTA radix top-k kernels keep histograms and inter-CTA barrier counters in `RadixRowState` scratch. The Python entry points previously obtained that scratch from device-keyed `_get_cache_buf` entries, so unsynchronized launches on different CUDA streams could share the same state and corrupt results or wedge the barrier.

The original PR covered seven producers. Current main subsequently added two supported `top_k_varlen` producers (`radix` and `radix_cutlass`), so this rebuild covers all nine current Python call sites.

## Fix

Allocate zero-initialized row-state scratch per call with `torch.zeros` instead of sharing a device cache entry. PyTorch's stream-ordered caching allocator gives concurrent live launches disjoint storage, reuses it after stream work completes, and handles CUDA graph capture without a custom ring or lifetime registry.

The `top_k_varlen` radix backend preserves its existing SM-count-based allocation size and `int32` view shape; the other eight paths retain their existing 1 MiB byte-buffer contract.

## Validation

Validated on an NVIDIA RTX PRO 6000 Blackwell, SM120, 96GB with CUDA 13:

- concurrent public API regressions: 5 passed (`top_k`, page-table transform, ragged transform, and both `top_k_varlen` backends)
- public-API pointer probe: distinct row-state allocations on two streams for both `top_k_varlen` backends
- focused sampling/logits selection: 1,104 passed, 123 skipped
- focused top-k/page-table/ragged selection: 163 passed
- exact-file pre-commit, compileall, and `git diff --check`: passed
- fresh independent outcome verification: confirmed

The earlier maintainer-triggered GPU pipeline `57443299` also succeeded on the previous converged per-call-allocation head. The current push only rebuilds that design on current main and extends it to the two newly landed producers.

## Disclosure

AI-assisted (Codex); reviewed and validated on real SM120 hardware by the submitter.

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved top-k and variable-length top-k operations during concurrent CUDA stream execution.
  - Prevented overlapping operations from sharing temporary state and producing incorrect results.
  - Preserved synchronization correctness across supported top-k, page-table, ragged, and radix-based processing modes.
  - Improved reliability when multiple sampling and probability-processing operations run concurrently.

- **Tests**
  - Added regression coverage for concurrent top-k and variable-length top-k operations across multiple CUDA streams.
  - Verified results for standard, page-table, ragged, and radix-based processing paths.

### CODE REVIEW

File: flashinfer/utils.py [hunk 1]
```diff
@@ -223,6 +224,86 @@ def _get_cache_buf(
...
+        buf = torch.zeros(bytes, dtype=torch.uint8, device=device)
+        _graph_capture_bufs.append(buf)
+        return buf
+    key = (name, device)
+    stream = torch.cuda.current_stream(device).cuda_stream
+    with _ring_cache_lock:
+        ring = _ring_cache_buf.get(key)
+        if ring is None or ring.slot_bytes < bytes:
+            ring = _RingBuf(bytes, num_slots, device)
+            _ring_cache_buf[key] = ring
```
[the remark this answers is not in the document]
Comment (waynehacking8) [reply flashinfer/utils.py]: Done in 847c439 — the cached ring is now rebuilt when a caller requests more slots (`ring.num_slots < num_slots`), alongside the existing slot-size growth check. Also renamed the builtin-shadowing `bytes` param to `nbytes` per the nitpick. 617 tests re-passed.

**Comment by waynehacking8** [comment]
@kahyunnam thanks for kicking off the pipeline. The GitLab run (#55035439) is internal, so I can't see which jobs failed — could you share the failing job names / a failed test, or point me at the log?

For context, I verified the changed multi-CTA top-k path locally on an RTX PRO 6000 (SM120):

```
pytest tests/utils/test_topk.py -k "multi_cta or refine_overflow or across_streams or dirty"
-> 102 passed, 0 failed
```

including the dirty-buffer-reuse and cross-stream determinism tests that exercise the new `_get_cache_buf_ring`. The two GitHub-side failures look unrelated to this change: **Deploy Docs** fails on `no module named flashinfer.comm` (a docs-build import; this PR doesn't touch `comm`, and `import flashinfer.comm` works fine here), and **Test Results Summary** is the gating job. So I suspect the internal failures are arch-specific (T4 / A10G / H100, which I can't run locally) or pre-existing flakes — but I'd rather confirm than assume. Which jobs went red?

**Comment by kahyunnam** [comment]
Hi @waynehacking8 the CI failures in gitlab logs seem unrelated.

File: flashinfer/utils.py [hunk 2]
```diff
@@ -223,6 +224,88 @@ def _get_cache_buf(
...
+_ring_cache_buf: Dict[Tuple[str, torch.device], _RingBuf] = {}
+_ring_cache_lock = threading.Lock()
+# Buffers handed out during CUDA graph capture. Their addresses are baked into
+# the captured graphs, so they must be kept alive for the lifetime of the
+# process (graphs may be replayed at any later time).
+_graph_capture_bufs: List[torch.Tensor] = []
+
+
+def _get_cache_buf_ring(
+    name: str, nbytes: int, device: torch.device, num_slots: int = 8
```
Comment (kahyunnam) [inline flashinfer/utils.py]: This = 8 is the only place the slot count is set. We might need a loud guard here when the buffer exceeds 8; otherwise this may also become a silent corruption issue

**Review by kahyunnam — COMMENTED** [review]
Thanks for the contribution! 

#3657 solved the same issue as #3618 bug (per-device scratch shared across multiple streams on same device) and fixed it with per-call allocation off PyTorch's stream-ordered caching allocator. What are your thoughts on converging this PR on that pattern, rather than carrying two solutions for one problem? 

It would also remove the silent-corruption-past-num_slots(8)-streams failure mode that may happen here.

**Comment by waynehacking8** [comment]
Thanks @kahyunnam — agreed, converged onto the #3657 pattern in 53849da.

`row_states` is now allocated per call straight from PyTorch's stream-ordered caching allocator (`torch.zeros(1MB, uint8, device=...)`) at all 7 call sites — sampling `top_k_renorm_probs`/`top_k_mask_logits`, topk `top_k`/`top_k_page_table_transform`/`top_k_ragged_transform`, and the two logits-processor ops. The caching allocator already hands each concurrent stream a disjoint block and recycles freed ones cheaply, so concurrent launches can't share the buffer (#3618), and it still works under CUDA-graph capture.

`_get_cache_buf_ring` / `_RingBuf` are deleted, so `utils.py` returns to baseline — one solution for the problem, not two. That also removes the `>num_slots(8)`-streams silent-corruption mode you flagged: there's no slot count left to overflow.

I also swapped the cache-reuse test for a concurrent-streams regression test (`test_multi_cta_row_states_not_shared_across_streams`) that interleaves `top_k` / `page_table` / `ragged` launches on two streams without syncing between them, mirroring #3657's `test_mxfp8_grouped_quantize_concurrent_streams`.

Verified on an RTX PRO 6000 (SM120, CUDA 13): the new test (all 3 APIs), the existing cross-stream determinism test, the mixed-k persistent-loop sampling tests, and the logits-processor top-k ops all pass.

**Comment by waynehacking8** [comment]
Quick CI status: the full GPU suite is green — `JIT Unittest` H100 / T4 / A10G (×5) and the entire `AOT Build Import` matrix (8/8 reruns) all pass.

The two remaining red checks are unrelated infra, not this change:
- `AOT Build Import (arm64, cu130)` was a spot-instance preemption mid-build (`The runner has received a shutdown signal`); its rerun passed.
- `remove-label` is the PR-label-cleanup workflow hitting `Resource not accessible by integration` (a token-permissions error in the workflow).

**Review by kahyunnam — APPROVED** [review]

**Comment by kahyunnam** [comment]
Will merge after bot CI comes back clean.
